File: src/tracker/lock.js
// Lock state machine — the QR-style "fire after N stable frames" gate.
//
// Direct port of autocapture/tracker/lock.py.
//
// States:
//   SEARCHING  — no lock, looking for a quad
//   UNSTABLE   — quad detected but failing a stability check
//   ALMOST     — stable and good quality but not yet for N frames
//   LOCKED     — ready to fire
//   COOLDOWN   — just fired; suppress until cooldown expires

import { KalmanQuad } from "./kalman.js";
import { meanCornerDrift, quadIou } from "../detector/quad.js";

export const STATE_SEARCHING = "SEARCHING";
export const STATE_UNSTABLE = "UNSTABLE";
export const STATE_ALMOST = "ALMOST";
export const STATE_LOCKED = "LOCKED";
export const STATE_COOLDOWN = "COOLDOWN";

export class LockGate {
  constructor(lockCfg) {
    this.cfg = lockCfg;
    this.kf = new KalmanQuad();
    this.state = STATE_SEARCHING;
    this._cooldownUntilFrame = 0;
    this._frameIdx = 0;
    this._prevSmoothed = null;
  }

  /**
   * Feed a new detection result.
   * @param {number[]|null} rawQuad
   * @param {number} quality
   * @returns {[string, number[]|null, number]} (state, smoothed, quality)
   */
  update(rawQuad, quality) {
    this._frameIdx += 1;
    const [smoothed] = this.kf.update(rawQuad);

    // Cooldown: suppress until it expires
    if (this.state === STATE_COOLDOWN) {
      if (this._frameIdx < this._cooldownUntilFrame) {
        return [STATE_COOLDOWN, smoothed, quality];
      }
      // cooldown ended — start fresh
      this.state = STATE_SEARCHING;
      this.kf.reset();
    }

    if (smoothed === null) {
      this.state = STATE_SEARCHING;
      this._prevSmoothed = null;
      return [STATE_SEARCHING, null, quality];
    }

    // Quality release rule (matches the python implementation's behavior)
    if (this.state === STATE_LOCKED && quality < 0.3) {
      this.kf.lockedFrames = 0;
      this.state = STATE_UNSTABLE;
      return [STATE_UNSTABLE, smoothed, quality];
    }

    // Drift gate — tolerance is RELATIVE to quad size. At close range the
    // quad spans most of the frame and a millimetre of hand tremor moves
    // corners >20px; a fixed threshold resets the gate every few frames
    // (smoothed quad vanishes, lock never engages). 4% of the diagonal
    // ≈ the same hand steadiness at any distance; floor at cornerDistPx.
    let drift = 0;
    if (this._prevSmoothed !== null) {
      drift = meanCornerDrift(smoothed, this._prevSmoothed);
    }
    this._prevSmoothed = Array.from(smoothed);

    const diag = Math.hypot(smoothed[4] - smoothed[0], smoothed[5] - smoothed[1]);
    const driftTol = Math.max(this.cfg.cornerDistPx, 0.04 * diag);
    if (drift > driftTol) {
      this.kf.lockedFrames = 0;
      this.state = STATE_UNSTABLE;
      return [STATE_UNSTABLE, smoothed, quality];
    }

    if (quality < 0.5) {
      this.kf.lockedFrames = 0;
      this.state = STATE_UNSTABLE;
      return [STATE_UNSTABLE, smoothed, quality];
    }

    // OK — accumulate stable frames
    this.kf.lockedFrames += 1;
    if (this.kf.lockedFrames >= this.cfg.nStableFrames) {
      this.state = STATE_LOCKED;
      return [STATE_LOCKED, smoothed, quality];
    }

    this.state = STATE_ALMOST;
    return [STATE_ALMOST, smoothed, quality];
  }

  /** Record that a capture just happened. Enters COOLDOWN. */
  fire() {
    this.state = STATE_COOLDOWN;
    // Convert cooldown_ms to frames (assume 30 fps; matches python default)
    const cooldownFrames = Math.floor(this.cfg.cooldownMs / 1000.0 * 30);
    this._cooldownUntilFrame = this._frameIdx + cooldownFrames;
    this.kf.reset();
    this._prevSmoothed = null;
  }
}

File: autocapture/tracker/lock.py
"""Lock state machine — the QR-style "fire after N stable frames" gate.

States
------
SEARCHING   — no lock, looking for a quad
UNSTABLE    — quad detected but failing one of the stability checks
ALMOST      — quad stable and good quality but not yet for N frames
LOCKED      — ready to fire
FIRING      — capture has just happened; cooldown active
"""

from __future__ import annotations
import numpy as np
from .kalman import KalmanQuad
from ..detector.quad import mean_corner_drift, quad_iou


STATE_SEARCHING = "SEARCHING"
STATE_UNSTABLE = "UNSTABLE"
STATE_ALMOST = "ALMOST"
STATE_LOCKED = "LOCKED"
STATE_FIRING = "FIRING"
STATE_COOLDOWN = "COOLDOWN"


class LockGate:
    """Wraps a KalmanQuad + stability counter + cooldown timer.

    The gate returns one of:
        ("SEARCHING", None)
        ("UNSTABLE",  smoothed_quad)
        ("ALMOST",   smoothed_quad)
        ("LOCKED",   smoothed_quad)         — caller should fire here
        ("COOLDOWN", smoothed_quad)         — just fired; ignore until cooldown ends
    """

    def __init__(self, lock_cfg):
        self.cfg = lock_cfg
        self.kf = KalmanQuad()
        self.state = STATE_SEARCHING
        self._cooldown_until_frame = 0
        self._frame_idx = 0
        self._prev_smoothed: np.ndarray | None = None

    def update(
        self,
        raw_quad: np.ndarray | None,
        quality: float,
    ) -> tuple[str, np.ndarray | None, float]:
        """Feed a new detection result. Returns (state, smoothed, quality).

        ``state`` is one of the STATE_* constants. ``smoothed`` is the
        Kalman-smoothed quad (or None if no detection). ``quality`` is
        passed through.
        """
        self._frame_idx += 1
        smoothed, innovation, _ = self.kf.update(raw_quad)

        # Cooldown: suppress everything until it expires (frame-counted,
        # not wall-clock, so the simulator and a real camera at 30 fps
        # behave the same).
        if self.state == STATE_COOLDOWN:
            if self._frame_idx < self._cooldown_until_frame:
                return STATE_COOLDOWN, smoothed, quality
            # cooldown ended — start fresh
            self.state = STATE_SEARCHING
            self.kf.reset()

        if smoothed is None:
            self.state = STATE_SEARCHING
            self._prev_smoothed = None
            return STATE_SEARCHING, None, quality

        # Quality gate: release lock if quality collapses
        if (self.state == STATE_LOCKED and
                quality < self.cfg.quality_release_ratio * 0):  # disabled
            pass
        # Apply the real release rule against the capture threshold
        # (quality_release_ratio < 1.0 means the gate is stricter than
        # the capture threshold, so a quality drop in the locked state
        # should release the lock)
        if self.state == STATE_LOCKED and quality < 0.3:
            self.kf.locked_frames = 0
            self.state = STATE_UNSTABLE
            return STATE_UNSTABLE, smoothed, quality

        # Drift gate
        if self._prev_smoothed is not None:
            drift = mean_corner_drift(smoothed, self._prev_smoothed)
        else:
            drift = 0.0
        self._prev_smoothed = smoothed.copy()

        if drift > self.cfg.corner_dist_px:
            self.kf.locked_frames = 0
            self.state = STATE_UNSTABLE
            return STATE_UNSTABLE, smoothed, quality

        if quality < 0.5:
            self.kf.locked_frames = 0
            self.state = STATE_UNSTABLE
            return STATE_UNSTABLE, smoothed, quality

        # Quality OK and drift OK — accumulate stable frames
        self.kf.locked_frames += 1
        if self.kf.locked_frames >= self.cfg.n_stable_frames:
            self.state = STATE_LOCKED
            return STATE_LOCKED, smoothed, quality

        self.state = STATE_ALMOST
        return STATE_ALMOST, smoothed, quality

    def fire(self) -> None:
        """Record that a capture just happened. Enters COOLDOWN."""
        self.state = STATE_COOLDOWN
        # Convert cooldown_ms to frames. The pipeline runs at fps frames
        # per second, so cooldown_ms / 1000 * fps = frames.
        # For now we use 30 fps as a reasonable default; the cooldown
        # length in frames is what the lock config controls.
        cooldown_frames = int(self.cfg.cooldown_ms / 1000.0 * 30)
        self._cooldown_until_frame = self._frame_idx + cooldown_frames
        self.kf.reset()
        self._prev_smoothed = None
